### ISSUE TITLE
Scope for operation and fragment names

### ISSUE THREAD

**Issue by asfernandes** [graphql/graphql-spec#802]
First of all, I understand what this spec calls a document and that a document is a request containing operations and fragments.

But tools also use "projects" (say, a set of documents with reusable fragments) and currently there is a problem in the way they do it. I think the way to make it better is having something on the spec specifying how to work with projects.

Let me detail the problem:

There are tools like generators (apollo and graphql-code-generator) and extensions (Apollo GraphQL for VSCode).

When you have these tools in your project, they consider that every file containing graphql operations and fragments are in a single scope. That is problematic because developer then need to use unique prefixes for names.

One may want to use per-component names (and hence, duplicate across whole project) but that is not possible.

In an analogy with sass, it have `:local` scope so one can define things only for that specific file.

So I suggest that the spec suggest something that allows names to be public (over the project) or local (file scope).

**Comment by dotansimha** [comment]
Totally agree with this one. 
We are using `documents` for GraphQL operations (query, mutation, subscription) and fragments. And `schema` for the GraphQL schema. 
But this is wrong, since schema can also be represented with a "document" (`DocumentNode`). 

I was thinking to change `documents` to `operations` in a future version of codegen, but this doesn't include fragments. So I think we might need a new terminology here. 

`Document` = The technical way of storing AST as object.

`Schema` = Actual `GraphQLSchema`

`Operation` = query/mutation/subscription.

`Operation+Fragment` = ? (I was thinking about "executables" but actually fragment is not executable by it's own).


@danielrearden I think you were struggling recently with naming of those things, right? (https://twitter.com/ReardenQL/status/1330583757217214466)

**Comment by sungam3r** [comment]
It seems to me that you are trying to solve problems not at the level of the language specification, but at the level of organization (projects/files/etc) in one or another IDE. There is simply no place for such concepts in the specification. At least I just don't understand what is _public_ or _local_ at the specification level.

**Comment by asfernandes** [comment]
> It seems to me that you are trying to solve problems not at the level of the language specification, but at the level of organization (projects/files/etc) in one or another IDE. There is simply no place for such concepts in the specification. At least I just don't understand what is _public_ or _local_ at the specification level.

I agree with you, but if the spec does not address this (a real world problem) at least with a suggestion, I don't think this real application problem of the technology will happen in a good way.

**Comment by sungam3r** [comment]
> if the spec does not address this (a real world problem) 

The point is that the specification doesn't _have to solve all the problems_. The specification has a clearly defined area of responsibility. And this is good.

> at least with a suggestion

Well, maybe only as a suggestion. But on the other hand, the specification should _specify_ the language/behavior/etc, and not give general suggestions about IDEs or something like that.

**Comment by asfernandes** [comment]
I'm not asking for suggesting an IDE or an extension.

It's just about admit people use them and suggest how they should work to use GraphQL in applications.

After all the focus of GraphQL is much more about people write applications than submit individual standalone queries typed and sent to the server.

Tools already chosen a way - treat names as global names - and I'm saying it is not the best way for all cases.

Maybe perhaps the spec decides to do nothing, but I still think it's good to have this discussion here so interested parts could know about it and have a design.

Maybe tools could just consider names started with `_` to be private? If not a simple decision like this is chosen, then I doubt we can achieve the goal without a broad specification for that concept.

**Comment by sungam3r** [comment]
Good. It will not be superfluous to ask such a question. I just doubt that the spec WG would be willing to take responsibility for resolving such issues from the spec _point of view_.

**Comment by Quramy** [comment]
I think it's related to https://github.com/graphql/graphql-spec/issues/343 discussion. If we discuss "scope", we should think how to determine boundary of scope or how to describe whether the members of the scope are exposed or not.
